eip-0023.md

```diff
@@ -0,0 +1,584 @@
+               INPUTS(1).tokens.size > 0                 &&
+               INPUTS(1).tokens(0)._1 == updateNFT       && // can only update when update box is the 2nd input
+               output.R4[GroupElement].get == selfPubKey && // public key is preserved
+               output.value >= SELF.value                && // value preserved or increased
+               ! (output.R5[Any].isDefined)                 // no more registers; prevents box from being reused as a valid vote 
+  
+  val owner = proveDlog(selfPubKey)
+  
+  // unlike in collection, here we don't require spender to be one of the ballot token holders
+  isSimpleCopy && (owner || update)
+}
+```
+
+### Update Contract
+
+```
+{ // This box (update box):
+  // Registers empty 
+  // 
+  // ballot boxes (Inputs)
+  // R4 the pub key of voter [GroupElement] (not used here)
+  // R5 the creation height of this box [Int]
+  // R6 the value voted for [Coll[Byte]] (hash of the new pool box script)
+  // R7 the reward token id in new box 
+  // R8 the number of reward tokens in new box 
+
+  val poolNFT = fromBase64("RytLYlBlU2hWbVlxM3Q2dzl6JEMmRilKQE1jUWZUalc=") // TODO replace with actual 
+
+  val ballotTokenId = fromBase64("P0QoRy1LYVBkU2dWa1lwM3M2djl5JEImRSlIQE1iUWU=") // TODO replace with actual 
+
+  val minVotes = 6 // TODO replace with actual
+  
+  val poolIn = INPUTS(0) // pool box is 1st input
+  val poolOut = OUTPUTS(0) // copy of pool box is the 1st output
+
+  val updateBoxOut = OUTPUTS(1) // copy of this box is the 2nd output
+
+  // compute the hash of the pool output box. This should be the value voted for
+  val poolOutHash = blake2b256(poolOut.propositionBytes)
+  val rewardTokenId = poolOut.tokens(1)._1
+  val rewardAmt = poolOut.tokens(1)._2
+  
+  val validPoolIn = poolIn.tokens(0)._1 == poolNFT
+  
+  val validPoolOut = poolIn.tokens(0) == poolOut.tokens(0)                && // NFT preserved
+                     poolIn.creationInfo._1 == poolOut.creationInfo._1    && // creation height preserved
+                     poolIn.value == poolOut.value                        && // value preserved 
+                     poolIn.R4[Long] == poolOut.R4[Long]                  && // rate preserved  
+                     poolIn.R5[Int] == poolOut.R5[Int]                    && // counter preserved
+                     ! (poolOut.R6[Any].isDefined)
+
+  
+  val validUpdateOut = updateBoxOut.tokens == SELF.tokens                     &&
+                       updateBoxOut.propositionBytes == SELF.propositionBytes &&
+                       updateBoxOut.value >= SELF.value                       &&
+                       updateBoxOut.creationInfo._1 > SELF.creationInfo._1    &&
+                       ! (updateBoxOut.R4[Any].isDefined) 
+
+  def isValidBallot(b:Box) = if (b.tokens.size > 0) {
+    b.tokens(0)._1 == ballotTokenId       &&
+    b.R5[Int].get == SELF.creationInfo._1 && // ensure vote corresponds to this box by checking creation height
```

Comment:
@scalahub I cannot understand the reason behind this check. I mean, why require the ballots and check the creation height of the existing update box? Why pool hash check below is not enough to make sure ballots are voting for this change?

Reply:
Oh, I see it now.  Makes sense. Thank you!